### ISSUE TITLE
Weird errors saying 'Cannot operate on a close database' or 'Suspicious behavior'

### ISSUE THREAD

**Issue by Woseseltops** [Signbank/Global-signbank#441]
All over Signbank. They go away when you reload the page. Restarting apache or uwsgi doesn't help.

**Comment by susanodd** [comment]
I tried looking up this error message. It says it can happen if a connection is closed.
Based on examples of the error, I tried searching on .close() and .cursor() in signbank.
This is done in various backend django files and jquery.
The only front-end code I could find is:
javascript:window.close() inside of signbank/dictionary/templates/dictionary/update_video.html
win.close() inside of media/admin/js/admin/RelatedObjectLookups.js
os.close(tmp) inside of signbank/video/fields.py
connection.cursor() inside of signbank/video/management/commands/importvideo.py

**Comment by susanodd** [comment]
It's weird that the update_video permission is only found in morpheme_detail.html
but not in gloss_detail.html (change_dataset is used instead.)
Do you remember what sequence of things you were doing?
Maybe some permissions are checked and okay, but then other different permissions are checked.
(Maybe during the sequence of events, a permission got dropped or a diferent one was needed instead.)

**Comment by ocrasborn** [comment]
Merel just tells me she's having this all the time today as well. Occurring upon various actions (editing, searching, adding signs).

**Comment by susanodd** [comment]
Can you give an example sequence? I can't generate this bug.

**Comment by ocrasborn** [comment]
Yesterday afternoon I could barely do anything without getting the 'closed database' error, today no problem at all. Reloading SB didn't solve anything yesterday.

**Comment by Woseseltops** [comment]
I also noticed this @ocrasborn , and it's related to a change/discovery I made:
* In my role as Ponyland admin, I've been doing a check to see how various Ponyland applications are doing every 15 mins. Until last week, uptime has been great for Signbank, but since then it fails seemingly at random.
* When Merel complained, the problems were increasing quickly; more fails than successes even.
* Out of pure desperation, I restarted Apache and uwsgi again... and then the failure frequency dropped. Seem like...

> Restarting apache or uwsgi doesn't help.

... isn't (completely) true.

This must mean that this is probably not a Signbank issue, but a Apache/uwsgi one. Similar to the ASL Signbank problems we had before, it seems like some problem is slowly building up, and you can 'empty' something by restarting. I'm happy with this information, because it means we have a workaround at least.

In case this problem persists, we could also consider not solving it and use the opportunity to migrate to the new Applejack. I don't expect a lot of problems (like with the ASL migration), as barely anything will change except from the actual hardware.

**Comment by ocrasborn** [comment]
Sounds like a good plan to me, @Woseseltops . Usage will be gradually increasing, with more datasets being created and new colleagues in the team who will work with Signbank.

**Comment by Woseseltops** [comment]
Relevant: I now restart when the automated check starts complaining. Has happened two times so far. I'll close this problem and start another one for the migration. #448

**Comment by Woseseltops** [comment]
Okay I've noticed this multiple times for the Global Signbank over the last few weeks, and now we also get this question from Julie for the ASL Signbank. Since this issue keeps appearing and then disappearing again, and because this issue is SQLite related, perhaps we should go for a 'deeper' solution and migrate to MySQL: #673

Fun fact: I think when we were talking about this in 2018, we did not realize this is a duplicate of issue #335 .

**Comment by Woseseltops** [comment]
Norwegian users of the Global Signbank also complain about this!

**Comment by susanodd** [comment]
This was happening in the tests when @TobiasCunnen tests were merged. There was a (test) folder created and then later deleted, but it wasn't necessarily completing the test because the tests get interleaved. So the folder was waiting to be deleted. (Something like that.) (All the tests were generating the locked database error.)

It seems large videos also time out (as suggested by @Woseseltops this is true) so the transaction doesn't complete. That was happening last week with Other Media. It was uploading okay, but then didn't get to the point of saving it in the database because the form wasn't available anymore. The code has been rewritten to create the object before the file was copied, and then delete the object if it failed to upload.

Perhaps this issue can be closed?

**Comment by susanodd** [comment]
@Woseseltops I'm getting this error a lot right now on Global!
I've searched on PIZZA with all the datasets selected.
I'm browsing each on via the sliding results bar in order to make sure the videos are visible.
(I've corrected two so far.)
It's awesome to see all the signs for pizza!

But three times now, this error has come up:


```
The above exception (database is locked) was the direct cause of the following exception:
/var/www/env/lib/python3.6/site-packages/django/contrib/sessions/backends/db.py in save

                    obj.save(force_insert=must_create, force_update=not must_create, using=using)
```

This is weird because I'm not editing anything, just browsing.
The operation "save" is somehow being called.

**Comment by susanodd** [comment]
Are large videos somehow causing this?
The ones I'm fixing happen to have the wrong permissions because they are large. (See recent issues related to uploading videos.) Perhaps the functioning of retrieving signs with large videos is causing a timeout. Although that "save" in the error message looks wrong. (De transactie is verkeerd bezig.)

**Comment by vanlummelhuizen** [comment]
@Woseseltops @susanodd This should be solved once we start using MySQL (see #673), right? So, close this issue?

**Comment by susanodd** [comment]
Sure, okay.
